src/test/mina_automation/daemon.ml

```diff
@@ -0,0 +1,186 @@
+    @param retry_delay The delay between retries.
+    @param retry_attempts The number of retries.
+    @return A deferred result indicating the success or failure of the operation. 
+  *)
+  let wait_for_bootstrap t ?(client_delay = 40.) ?(retry_delay = 60.)
+      ?(retry_attempts = 10) () =
+    Async.printf "Waiting initial %d s. before connecting\n"
+      (int_of_float client_delay) ;
+    let%bind _ =
+      Deferred.map (after @@ Time.Span.of_sec client_delay) ~f:Or_error.return
+    in
+    let rec go retries_remaining =
+      let%bind output = daemon_status t in
+      Async.printf "%s" output ;
+      if String.is_substring output ~substring:"Synced" then
+        Deferred.Or_error.ok_unit
+      else if retries_remaining > 0 then (
+        Async.printf "Daemon not responding.. retrying (%i/%i)\n"
+          (retry_attempts - retries_remaining)
+          retry_attempts ;
+        let%bind () = after @@ Time.Span.of_sec retry_delay in
+        go (retries_remaining - 1) )
+      else Deferred.Or_error.error_string output
+    in
+    go retry_attempts
+end
+
+module Config = struct
+  module ConfigDirs = struct
+    type t =
+      { root_path : string
+      ; conf : Filename.t
+      ; genesis : Filename.t
+      ; libp2p_keypair : Filename.t
+      }
+
+    let libp2p_keypair_folder t = String.concat [ t.libp2p_keypair; "/privkey" ]
+
+    let create ?(root_path = "/tmp") ?(config_dir = "mina_spun_test")
+        ?(genesis_dir = "mina_genesis_state")
+        ?(p2p_dir = "mina_test_libp2p_keypair") () =
+      (* create empty config dir to avoid any issues with the default config dir *)
+      let conf = Filename.temp_dir ~in_dir:root_path config_dir "" in
+      let genesis = Filename.temp_dir ~in_dir:root_path genesis_dir "" in
+      let libp2p_keypair = Filename.temp_dir ~in_dir:root_path p2p_dir "" in
+      { root_path; conf; genesis; libp2p_keypair }
+
+    let dirs t = [ t.conf; t.genesis; t.libp2p_keypair ]
+
+    let mina_log t = t.conf ^/ "mina.log"
+  end
+
+  type t = { port : int; dirs : ConfigDirs.t }
+
+  let default ?dirs =
+    { port = 3085
+    ; dirs = (match dirs with Some dirs -> dirs | None -> ConfigDirs.create ())
+    }
+
+  let libp2p_keypair_folder t = ConfigDirs.libp2p_keypair_folder t.dirs
+
+  let generate_keys t =
+    let open Deferred.Let_syntax in
+    let%map () =
+      Init.Client.generate_libp2p_keypair_do (libp2p_keypair_folder t) ()
+    in
+    ()
+end
+
+(** 
+  Module [Process] provides functions to interact with a Mina daemon process.
+*)
+module Process = struct
+  type t = { process : Process.t; config : Config.t; client : Client.t }
+
+  (** [force_kill t] sends a kill signal to the process associated with [t] and waits for the process to terminate.
+    @param t The daemon instance containing the process to be killed.
+    @return A deferred result indicating the success or failure of the operation.
+  *)
+  let force_kill t =
```

Comment:
These functions occurs a lot across different Process module. I think this is a nice place to actually use an OOP/Inheritance pattern.